enterprise/coderd/identityprovider/tokens.go

```diff
@@ -0,0 +1,246 @@
+	p := httpapi.NewQueryParamParser()
+	err := r.ParseForm()
+	if err != nil {
+		return tokenParams{}, nil, xerrors.Errorf("parse form: %w", err)
+	}
+
+	// TODO: Can we make this a URL straight out of the database?
+	cb, err := url.Parse(callbackURL)
+	if err != nil {
+		return tokenParams{}, nil, err
+	}
+
+	p.Required("grant_type", "client_secret", "client_id", "code")
+
+	vals := r.Form
+	params := tokenParams{
+		clientID:     p.String(vals, "", "client_id"),
+		clientSecret: p.String(vals, "", "client_secret"),
+		code:         p.String(vals, "", "code"),
+		redirectURL:  p.URL(vals, cb, "redirect_uri"),
+		grantType:    httpapi.ParseCustom(p, vals, "", "grant_type", httpapi.ParseEnum[codersdk.OAuth2ProviderGrantType]),
+	}
+
+	p.ErrorExcessParams(vals)
+	return params, p.Errors, nil
+}
+
+func Tokens(db database.Store, defaultLifetime time.Duration) http.HandlerFunc {
+	return func(rw http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		app := httpmw.OAuth2ProviderApp(r)
+
+		params, validationErrs, err := extractTokenParams(r, app.CallbackURL)
+		if err != nil {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "Failed to validate form values.",
+				Detail:  err.Error(),
+			})
+			return
+		}
+		if len(validationErrs) > 0 {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message:     "Invalid query params.",
+				Validations: validationErrs,
+			})
+			return
+		}
+
+		var token oauth2.Token
+		//nolint:gocritic,revive // More cases will be added later.
+		switch params.grantType {
+		// TODO: Client creds, device code, refresh.
+		default:
+			token, err = authorizationCodeGrant(ctx, db, app, defaultLifetime, params.clientSecret, params.code)
+		}
+
+		if err != nil && errors.Is(err, sql.ErrNoRows) {
+			httpapi.Write(r.Context(), rw, http.StatusUnauthorized, codersdk.Response{
+				Message: "Invalid client secret or code",
+			})
+			return
+		}
+		if err != nil {
+			httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "Failed to exchange token",
+				Detail:  err.Error(),
+			})
+			return
+		}
+
+		// Some client libraries allow this to be "application/x-www-form-urlencoded". We can implement that upon
+		// request. The same libraries should also accept JSON. If implemented, choose based on "Accept" header.
+		httpapi.Write(ctx, rw, http.StatusOK, token)
+	}
+}
+
+func authorizationCodeGrant(ctx context.Context, db database.Store, app database.OAuth2ProviderApp, defaultLifetime time.Duration, clientSecret, code string) (oauth2.Token, error) {
+	// Validate the client secret.
+	secretHash := Hash(clientSecret, app.ID)
+	secret, err := db.GetOAuth2ProviderAppSecretByAppIDAndSecret(
+		//nolint:gocritic // Users cannot read secrets so we must use the system.
+		dbauthz.AsSystemRestricted(ctx),
+		database.GetOAuth2ProviderAppSecretByAppIDAndSecretParams{
+			AppID:        app.ID,
+			HashedSecret: secretHash[:],
+		})
```

Comment:
Ooooh I now see why you did this. My above comment about using `userpassword.Hash` remains, I think we should just prefix this stuff with some id and use `userpassword.Compare(hashedSecret, secret)` to validate.

This would mean you would have to store an extra id in the database...

So currently a secret is 40 random characters. We should prefix it with something.

Before: `EA9ILdVv0wVgXo840dFZV4UX21dtZsl5rziDiiU7`

New: `coder_io928hVamL_EA9ILdVv0wVgXo840dFZV4UX21dtZsl5rziDiiU7` (or something like this).

You would have to store `io928hVamL` in the db, so when we are given the secret, we split on `_`, find the `id` and do the search with that. This column also needs a `unique` constraint.